src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/vmware/VMWareProviderVirtualMachineUtil.java

```diff
@@ -64,91 +66,128 @@
+                        .filter(virtualMachine -> virtualMachine.getConfig().getUuid().equals(uuid))
+                        .findFirst();
+        } catch (RemoteException e) {
+            throw new RuntimeException("ERROR when retrieving VMWare virtual machines" + uuid, e);
+        }
     }
 
-    public Set<VirtualMachine> getAllVirtualMachinesByInfrastructure(Folder rootFolder, Infrastructure infrastructure) {
+    public Set<VirtualMachine> getAllVirtualMachines(Folder rootFolder) {
         try {
-
             ManagedEntity[] managedEntities = new InventoryNavigator(rootFolder).searchManagedEntities(EntityType.VM.getValue());
 
             return IntStream.range(0, managedEntities.length)
                             .mapToObj(i -> (VirtualMachine) managedEntities[i])
                             .collect(Collectors.toSet());
 
         } catch (RemoteException e) {
-            throw new RuntimeException("ERROR when retrieving VMWare instances for infrastructure : " + infrastructure,
-                                       e);
+            throw new RuntimeException("ERROR when retrieving VMWare virtual machines", e);
         }
     }
 
-    public Folder searchFolderByName(String name, Folder rootFolder) throws RemoteException {
-        return (Folder) new InventoryNavigator(rootFolder).searchManagedEntity(EntityType.FOLDER.getValue(), name);
+    public Optional<HostSystem> searchHostByName(String name, Folder rootFolder) {
+        try {
+            return Optional.ofNullable((HostSystem) new InventoryNavigator(rootFolder).searchManagedEntity(EntityType.HOST.getValue(),
+                                                                                                           name));
+        } catch (RemoteException e) {
+            throw new RuntimeException("ERROR when retrieving VMWare host with name: " + name, e);
+        }
     }
 
-    public HostSystem searchHostByName(String name, Folder rootFolder) throws RemoteException {
-        return (HostSystem) new InventoryNavigator(rootFolder).searchManagedEntity(EntityType.HOST.getValue(), name);
+    public Optional<Folder> searchFolderByName(String name, Folder rootFolder) {
+        try {
+            return Optional.ofNullable((Folder) new InventoryNavigator(rootFolder).searchManagedEntity("Folder", name));
+        } catch (RemoteException e) {
+            throw new RuntimeException("ERROR when retrieving VMWare folder with name: " + name, e);
+        }
     }
 
-    public Folder searchVMFolderFromVMName(String name, Folder rootFolder) throws RemoteException {
-        ManagedEntity current = searchVirtualMachineByName(name, rootFolder).getParent();
-        while (!(current instanceof Folder)) {
-            current = current.getParent();
-        }
-        return (Folder) current;
+    public Optional<Folder> searchVMFolderFromVMName(String name, Folder rootFolder) {
+        final ManagedEntity[] vmFolder = new Folder[1];
```

Comment:
Because finals are required by the lambda expression `ifPresent()`, intellij proposed me the array trick. Anyway I replaced it by a regular `if(isPresent())` in 4d4e5a0c0ddcf458af8f3d3e46e207e7a99e4ad3